### PR TITLE
PI-based modeling of controller Packet-IO

### PR DESCRIPTION
**Pull request by wmohsin**

### CODE REVIEW

File: proto/p4/config/p4info.proto [hunk 1]
```diff
@@ -197,3 +198,19 @@ message DirectMeter {
   // the id of the table to which the meter is attached
   uint32 direct_table_id = 3;
 }
+
+// packet-in & packet-out metadata (consistent with p4runtime.proto)
+message ControllerPacketMetadata {
```
Comment (teverman) [inline proto/p4/config/p4info.proto]: For my info, in p4c, the P4 runtime serializer is expected to add this to the P4Info based on header definitions?

File: proto/p4/config/p4info.proto [hunk 2]
```diff
@@ -197,3 +198,19 @@ message DirectMeter {
...
+  // preamble.name and preamble.id will specify header type (“packet_out” or
+  // “packet_in” for now).
+  Preamble preamble = 1;
+  message Metadata {
+    uint32 id = 1;
+    string name = 2;
```
Comment (teverman) [inline proto/p4/config/p4info.proto]: Is this a fully-qualified name? Please clarify with comment.

File: proto/p4/p4runtime.proto [hunk 3]
```diff
@@ -334,8 +342,20 @@ message StreamMessageResponse {
...
+  repeated PacketMetadata metadata = 2;
+
+  // TODO: Remove fields below once users of PacketIn have migrated to
+  // PacketMetadata.
+  uint32 ingress_physical_port = 3;
+  uint32 ingress_logical_port = 4;
+}
+
+message PacketMetadata {
+  uint32 metadata_id = 1;
```
Comment (teverman) [inline proto/p4/p4runtime.proto]: Please add a comment to explain the metadata_id.  Does it refer to the ID in the P4Info metadata fields in ControllerPacketMetadata?

File: proto/p4/config/p4info.proto (hunk 1 above)
Comment (antoninbas) [inline proto/p4/config/p4info.proto]: yes, based on annotations on the header definitions

File: proto/p4/config/p4info.proto (hunk 2 above)
Comment (wmohsin) [inline proto/p4/config/p4info.proto]: Done.

**Pull request merged**